lttngpy/src/lttngpy/channel.cpp

```diff
@@ -0,0 +1,87 @@
+// Copyright 2023 Apex.AI, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <lttng/channel.h>
+#include <lttng/constant.h>
+#include <lttng/domain.h>
+#include <lttng/event.h>
+#include <lttng/handle.h>
+#include <lttng/lttng-error.h>
+
+#include <optional>
+#include <string>
+
+#include "lttngpy/channel.hpp"
+
+namespace lttngpy
+{
+
+int enable_channel(
+  const std::string & session_name,
+  const enum lttng_domain_type domain_type,
+  const enum lttng_buffer_type buffer_type,
+  const std::string & channel_name,
+  const std::optional<int> overwrite,
+  const std::optional<uint64_t> subbuf_size,
+  const std::optional<uint64_t> num_subbuf,
+  const std::optional<uint64_t> switch_timer_interval,
+  const std::optional<uint64_t> read_timer_interval,
+  const std::optional<enum lttng_event_output> output)
+{
+  struct lttng_domain domain {};
+  domain.type = domain_type;
+  domain.buf_type = buffer_type;
+
+  struct lttng_handle * handle = lttng_create_handle(session_name.c_str(), &domain);
+  if (nullptr == handle) {
+    return -LTTNG_ERR_UNK;
```

Comment:
Just out of curiosity, why negative here?

Reply:
That's the behaviour of the `lttng_enable_channel()` function that this wraps: https://github.com/lttng/lttng-tools/blob/d4333eb996dec315fb721074c2cf49adee04363b/include/lttng/channel.h#L87. The documentation for `lttngpy::enable_channel()` says the same thing: https://github.com/ros2/ros2_tracing/blob/ca833252f8e35bf86c0da7ff7e68678ed796b36e/lttngpy/src/lttngpy/channel.hpp#L41

The lttng-ctl error code constants are positive: https://github.com/lttng/lttng-tools/blob/d4333eb996dec315fb721074c2cf49adee04363b/include/lttng/lttng-error.h#L24-L217. When used as a return value, they are negated. A positive return value means that the call was successful, see for example `lttng_list_channels()`: https://github.com/lttng/lttng-tools/blob/d4333eb996dec315fb721074c2cf49adee04363b/include/lttng/channel.h#L77C74-L77C74. `lttng_strerror()` can handle both negative and positive error codes, though.